poll GPU status on initial installation

cmluciano opened the pull request:
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: related #44080

**Release note**:
```
retry GPU card discovery in a goroutine upon startup
```

Commit:
Detect expected mounted GPUs before discovering

If nvidia drivers have not been properly installed, some video cards
will not appear under /dev as mountable.

We now check the count of devices expected to be found under dev in
 the devices procfile. Cards will not be added to the allocatable pool
 unless expected cards matches found cards.

Closes #44080

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44463)
<!-- Reviewable:end -->

cmluciano commented:
cc @vishh

cmluciano commented:
@vishh for retry logic, is https://github.com/kubernetes/kubernetes/tree/master/pkg/util/goroutinemap preferred?

vishh commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -97,8 +98,8 @@ func (ngm *nvidiaGPUManager) Start() error {
-
-	if err := ngm.discoverGPUs(); err != nil {
+	
+	if err := ngm.discoverGPUs("/proc/bus/pci/devices", "/dev"); err != nil {
```
nit: Would prefer constants for these strings

vishh replied:
Let's not fail Start() if gpu discovery fails initially. Instead log discovery errors and try again. You can start a go thread that will continue to discover GPUs until it succeeds.

```go
go func() {
  var err error
  for {
    if err := ngm.discoverGPUs(...); err == nil {
       return
   }
   glog.Errorf("failed to discover Nvidia GPUs: %v", err)
   time.Sleep(time.Minute)
 }
}()
```

vishh commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -190,9 +191,14 @@ func (ngm *nvidiaGPUManager) updateAllocatedGPUs() {
+	}
 	reg := regexp.MustCompile(nvidiaDeviceRE)
 	files, err := ioutil.ReadDir(devDirectory)
+	foundGPUs := sets.NewString()
```
nit: Move this declaration below the error evaluation

vishh commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -202,13 +208,44 @@ func (ngm *nvidiaGPUManager) discoverGPUs() error {
 
+// expectedGPUs reads the devices procfile and determines how many GPUs we should find
+// when discoverGPUs is called
+// TODO: This most likely will count dead GPUs and should be refactored if we
```
I'd recommend opening an issue for this and adding it to our tracker issue.

cmluciano replied:
Linked in my NVML proposal for now. I can split that proposal up to just suit the general use case of discovery/health-checking.

vishh commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -202,13 +208,44 @@ func (ngm *nvidiaGPUManager) discoverGPUs() error {
+	for scanner.Scan() {
+		s := scanner.Bytes()
+		fields := bytes.SplitN(s, []byte{'\t'}, 3)
+		if bytes.HasPrefix(fields[1], []byte("10de")) {
```
make `10de` a const with more documentation

grodrigues3 commented:
@k8s-bot ok to test

flx42 commented:
Is this only to solve the issue in #44080, i.e. a concurrent driver installation?
If that's the case, it seems highly dangerous, for little added value. I'm not sure this is the right place to solve this.

BTW, I'm pretty sure your code is incorrect for this case:
```
$ lspci -d 10de:                                                                                                                                                                                                                                                           
01:00.0 VGA compatible controller: NVIDIA Corporation GM204 [GeForce GTX 980] (rev a1)                                                                                                                                                                                         
01:00.1 Audio device: NVIDIA Corporation GM204 High Definition Audio Controller (rev a1)                                                                                                                                                                                       
02:00.0 VGA compatible controller: NVIDIA Corporation Device 1b80 (rev a1)                                                                                                                                                                                                     
02:00.1 Audio device: NVIDIA Corporation Device 10f0 (rev a1) 
```
You will return `4`, right?

Commit:
Continually poll for GPU drivers on startup

This gorounine allows  parrallel and serial installations of GPU drivers in
tandem with kubelet setup.

It will continue to look for entries in /dev if it detects GPUs exist
in the devices procfile.

Related #44080

cmluciano commented:
@flx42 I do believe that may be possible if the devices are plugged into the same PCI slots that GPUs would be.

I will try to grok more of this file to glean more metadata.

flx42 commented:
> I do believe that may be possible if the devices are plugged into the same PCI slots that GPUs would be.

What do you mean? There is *one* physical NVIDIA GPU board, plugged into *one* PCIe slot. Those two entries refer to the same board (they are also part of the same IOMMU group).
Those GPUs have an HDMI output so they need to expose a sound device.

This is just one example, I think there are many pitfalls here, I still think it's risky to try to implement this manually.

vishh commented:
Another option is to expose a knob that will whitelist specific device IDs
for now. We cannot add a nvml dependency to a patch release.

On Fri, Apr 14, 2017 at 10:37 AM, Felix Abecassis <notifications@github.com>
wrote:

> I do believe that may be possible if the devices are plugged into the same
> PCI slots that GPUs would be.
>
> What do you mean? There is *one* physical NVIDIA GPU board, plugged into
> *one* PCIe slot. Those two entries refer to the same board (they are also
> part of the same IOMMU group).
> Those GPUs have an HDMI output so they need to expose a sound device.
>
> This is just one example, I think there are many pitfalls here, I still
> think it's risky to try to implement this manually.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44463#issuecomment-294198470>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKEIpE5JYbTAWZfrYz8Qgij0xtOQsks5rv67XgaJpZM4M9Drh>
> .
>

cmluciano commented:
I'm not sure if that is feasible looking at the breadth of IDs listed [here](http://envytools.readthedocs.io/en/latest/hw/pciid.html#nv40)

vishh commented:
How about an small whitelist and have users update the whitelist?

On Fri, Apr 14, 2017 at 10:45 AM, cmluciano <notifications@github.com>
wrote:

> I'm not sure if that is feasible looking at the breadth of IDs listed here
> <http://envytools.readthedocs.io/en/latest/hw/pciid.html#nv40>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44463#issuecomment-294200043>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKNqrJjVZX5n3DDYfCF_kGx4jn9Daks5rv7CfgaJpZM4M9Drh>
> .
>

flx42 commented:
@vishh this sounds complicated, for what added value? I'm still not sure.
Is it only because the driver installation is done concurrently by some users?

vishh commented:
@flx42 this is to avoid a kubelet restart. The other option is to have the installer restart kubelet process which is also sub-optimal.

vishh commented:
@flx42 we are open to suggestion, so feel free to help out!

cmluciano commented:
@flx42 Does your grep print multiple lines for ``10de`` when the ``/proc/bus/pci/devices``` is targeted?

@vishh I can try to parse the output of ``lspci`` if that would be more effective

vishh commented:
On Mon, Apr 17, 2017 at 8:17 AM, cmluciano <notifications@github.com> wrote:

> @flx42 <https://github.com/flx42> Does your grep print multiple lines for
> 10de when the ``/proc/bus/pci/devices``` is targeted?
>
> @vishh <https://github.com/vishh> I can try to parse the output of lspci
> if that would be more effective
>
I was hoping to avoid yet another external binary dependency.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44463#issuecomment-294506859>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKPpoxx4oybWfd79x_2bKmgJtkB9wks5rw4J1gaJpZM4M9Drh>
> .
>

flx42 commented:
> Does your grep print multiple lines for 10de when the ``/proc/bus/pci/devices``` is targeted?

You mean like this?

```
$ lspci -nn | grep 10de
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation GM204 [GeForce GTX 980] [10de:13c0] (rev a1)
01:00.1 Audio device [0403]: NVIDIA Corporation GM204 High Definition Audio Controller [10de:0fbb] (rev a1)
02:00.0 VGA compatible controller [0300]: NVIDIA Corporation Device [10de:1b80] (rev a1)
02:00.1 Audio device [0403]: NVIDIA Corporation Device [10de:10f0] (rev a1)
```

cmluciano commented:
@flx42 Yes do these show up as separate entries in ``/proc/bus/pci/devices``

flx42 commented:
@cmluciano Yes.

cmluciano commented:
@vishh Any ideas on how to proceed? We missed the latest release cutoff for 1.6, so we have a bit more time to noodle on what we would like to try out.

vishh commented:
@cmluciano for now let's stick to restarting kubelet. Let's discuss this further during the upcoming f2f.

cmluciano commented:
@vishh Do we want to close this PR but keep the issue open or iterate on this PR?

vishh commented:
SGTM

On Tue, May 23, 2017 at 2:33 AM, cmluciano <notifications@github.com> wrote:

> @vishh <https://github.com/vishh> Do we want to close this PR but keep
> the issue open or iterate on this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44463#issuecomment-303152379>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKDn1kCDcgDOQGjJKg_BoVRFs8D0mks5r8bjZgaJpZM4M9Drh>
> .
>

cmluciano commented:
Reopening  for comment. 

@vishh In the [e2e tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/gpus.go#L63) you are looking under ``/sys/devices/pci-*`` for NVIDIA entries. Did you find this location to be more optimal for detecting the number of expected cards?

vishh commented:
In the e2es all I need is to identify if **any** Nvidia GPUs exist on a given node. I do not need to know the exact number which IIUC is what this PR is trying to tackle.

cmluciano commented:
@vishh Do you think it is sufficient in this PR to do the same and only detect if we should expect GPUs or not?

vishh commented:
Yeah. That might be a good idea.

On Mon, Jun 5, 2017 at 8:21 AM, cmluciano <notifications@github.com> wrote:

> @vishh <https://github.com/vishh> Do you think it is sufficient in this
> PR to do the same and only detect if we should expect GPUs or not?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44463#issuecomment-306215992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKGo_DPcTeHTz893rV9rlhP0IGxSdks5sBBzjgaJpZM4M9Drh>
> .
>

cmluciano commented:
@vishh updated

cmluciano commented:
/retest

cmluciano commented:
@vishh Tests are passing

cc @mindprince for a review

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -98,11 +104,19 @@ func (ngm *nvidiaGPUManager) Start() error {
+	go func() {
+		var err error
+		for {
+			if err := ngm.discoverGPUs(devicesProcFile, devDirectory); err == nil {
```
`err` should not be redeclared here, otherwise the log statement below will always show `nil` error.

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -190,25 +204,61 @@ func (ngm *nvidiaGPUManager) updateAllocatedGPUs() {
+		if bytes.HasPrefix(fields[1], []byte(nvidiaVendorID)) {
+			return true, nil
+		}
+
```
Nit: extra newline.

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -98,11 +104,19 @@ func (ngm *nvidiaGPUManager) Start() error {
+			if err := ngm.discoverGPUs(devicesProcFile, devDirectory); err == nil {
+				return
+			}
+			glog.Errorf("failed to discover Nvidia GPUs: %v", err)
```
This can potentially cause a flood of unending error messages in the log file.

cmluciano replied:
I'm worried about this as well. Should it be changed to an info message or should I reconsider whether to log at all?

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -190,25 +204,61 @@ func (ngm *nvidiaGPUManager) updateAllocatedGPUs() {
 // name.
-func (ngm *nvidiaGPUManager) discoverGPUs() error {
+func (ngm *nvidiaGPUManager) discoverGPUs(devicesFile string, devDir string) error {
+	found, err := expectedGPUs(devicesFile)
```
Nit: call this `expected` instead of `found`?

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager_test.go:
```diff
@@ -36,6 +39,52 @@ func (tapl *testActivePodsLister) GetActivePods() []*v1.Pod {
+	content := []byte("aaa\t10debbb\tccc\nddd\teee\tfff\nggg\t10dehhh\tiii\n")
+	tmpfile, err := ioutil.TempFile("", "tmp")
+	if err != nil {
+		fmt.Print(err)
```
Why are we continuing after error?

cmluciano replied:
Forgot to flip this after debugging :P

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager_test.go:
```diff
@@ -36,6 +39,52 @@ func (tapl *testActivePodsLister) GetActivePods() []*v1.Pod {
+	content := []byte("dummy content")
+	dir, err := ioutil.TempDir("", "tmpdir")
+	if err != nil {
+		fmt.Print(err)
```
Same here, we shouldn't continue after error.

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager_test.go:
```diff
@@ -36,6 +39,52 @@ func (tapl *testActivePodsLister) GetActivePods() []*v1.Pod {
+	return dir
+}
+
+func TestDiscoverGPUS(t *testing.T) {
```
Can you add some more test cases?
- To test behavior when no expected devices.
- To test behavior when there are expected devices but no found devices.

cmluciano replied:
👍🏽

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -190,25 +204,61 @@ func (ngm *nvidiaGPUManager) updateAllocatedGPUs() {
+func (ngm *nvidiaGPUManager) discoverGPUs(devicesFile string, devDir string) error {
+	found, err := expectedGPUs(devicesFile)
+	if err != nil || !found {
+		return err
```
It was not directly obvious to me that this could be a return path in normal (non error circumstances). I would recommend splitting this if condition into two blocks, one for `err != nil` and another for the other non-error case.

cmluciano commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -98,11 +104,19 @@ func (ngm *nvidiaGPUManager) Start() error {
+		}
+	}()
+	// It's possible that the runtime isn't available now.
+	ngm.allocated = ngm.gpusInUse()
```
I think this was a bad rebase on my part.

cmluciano commented:
/retest

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager_test.go:
```diff
@@ -36,6 +38,72 @@ func (tapl *testActivePodsLister) GetActivePods() []*v1.Pod {
+	noFoundGPUs, err := expectedGPUs(testProc)
+	as := assert.New(t)
+	as.Nil(err)
+	as.Equal(noFoundGPUs, false)
```
Also test that `discoverGPUs` doesn't return an error in this case. Would have caught the bug I pointed out above.

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager_test.go:
```diff
@@ -36,6 +38,72 @@ func (tapl *testActivePodsLister) GetActivePods() []*v1.Pod {
+		allGPUs: sets.NewString(),
+	}
+	err = testGpuManager.discoverGPUs(testProc, testDir)
+	as.NotNil(err, "expected to find GPUs but not find them under the deviceDir: %v", testDir)
```
Since this error message is for the test run, it should say something like, "expected an error because of not finding GPUs".

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -98,11 +105,15 @@ func (ngm *nvidiaGPUManager) Start() error {
+	go func() {
+		for {
+			if found := ngm.discoverGPUs(devicesProcFile, devDirectory); found == nil {
+				return
```
We should log that GPU discover finished, so that there's an easy way to find out whether this goroutine is running or finished.

mindprince commented on pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go:
```diff
@@ -98,11 +104,16 @@ func (ngm *nvidiaGPUManager) Start() error {
+			if err := ngm.discoverGPUs(devicesProcFile, devDirectory); err == nil {
+				return
+			}
+			glog.V(4).Infof("Nvidia GPU discovery finished")
```
This should be inside the `if` condition before the `return` statement.

cmluciano replied:
nice catch :)

rohitagarwal003 reviewed (changes requested):
One minor change, everything else looks good.

mindprince commented:
/assign

mindprince commented:
/approve

mindprince commented:
/retest

mindprince commented:
/retest

cmluciano commented:
I think we can wait on this for 1.8

cmluciano commented:
@mindprince Can I get a LGTM on this?

mindprince commented:
I am not in the OWNERS file yet, my LGTM wouldn't help. :)

Also, I was wondering if you ran the e2e gpu tests on this? It looks good but I would like to have an e2e run with gpus enabled to be sure. I want to get #47192 done soon, so that this uncertainty goes away.

cmluciano commented:
Yes perhaps it is best to await that PR landing first.

cmluciano commented:
/retest

cmluciano commented:
/retest

federation suite is fixed

cmluciano commented:
@vishh @mindprince Are we running e2es for GPU on every pull now?

cmluciano commented:
/retest

cmluciano commented:
/test pull-kubernetes-e2e-gce-gpu

cmluciano commented:
/test pull-kubernetes-e2e-gce-gpu

cmluciano commented:
/retest

cluster missing 1 node

cmluciano commented:
/test pull-kubernetes-e2e-gce-gpu

cmluciano commented:
ping @mindprince is this GPU test running on this PR?

mindprince commented:
Yeah, it's passing: https://k8s-gubernator.appspot.com/pr/44463

cmluciano commented:
Sweet, ping @vishh for approval

vishh commented:
The current integration with Nvidia GPUs is planned to be deprecated in favor of device plugins. That said, I see this PR as a bug fix.
I'd like to hear @flx42's opinion on if this actually fixes the bug.

cmluciano commented:
@flx42 @RenaudWasTaken 

Any opinions on if we think this fixes the bug until the device plugin rebase?

RenaudWasTaken commented:
I think that if your intent is to know if there is at least one NVIDIA device that would work (@3XX0 and @flx42  would have more insights on this though).

Do we intend to merge this before 1.8 ?

cmluciano commented:
@vishh should this be considered a bug fix for 1.8?

RenaudWasTaken commented:
@cmluciano device plugin should land in 1.8

vishh commented:
I do consider this as a bug fix. However the old implementation is in the process of being deprecated in favor of device plugins. Should we take the risk of merging this PR and possibly destabilizing v1.8?
Personally, the risk outweighs the potential benefits. If we are merging this we should ideally back port the fix.

cmluciano commented:
@vishh I agree. I will close this in favor of the new functionality from device-plugins. I can add it there if we still want this.
